File: packages/python/src/ngtaxkit/invoice.py
"""Invoice module (Layer 2) — create, validate, and export invoices."""

from __future__ import annotations

import datetime
import json
from typing import Any

from . import vat as vat_module
from .errors import EmptyInvoiceError, InvalidAmountError, InvalidQuantityError
from .utils import bankers_round


def create(
    *,
    seller: dict[str, Any],
    buyer: dict[str, Any],
    items: list[dict[str, Any]],
    invoice_number: str,
    currency: str = "NGN",
    issue_date: str | None = None,
    due_date: str | None = None,
    purchase_order_ref: str | None = None,
    notes: str | None = None,
    type: str = "invoice",
) -> dict[str, Any]:
    """Create an invoice with auto-calculated VAT per line item."""
    if not items:
        raise EmptyInvoiceError("Invoice must have at least one line item")

    for i, item in enumerate(items):
        unit_price = item.get("unit_price", 0)
        if unit_price < 0:
            raise InvalidAmountError(f"Line item {i + 1} has negative unit price: {unit_price}")
        qty = item.get("quantity", 0)
        if qty <= 0:
            raise InvalidQuantityError(f"Line item {i + 1} has zero or negative quantity: {qty}")

    if issue_date is None:
        issue_date = datetime.date.today().isoformat()

    computed_items: list[dict[str, Any]] = []
    for item in items:
        category = item.get("category", "standard")
        unit_price = item.get("unit_price", 0)
        quantity = item["quantity"]
        line_net = bankers_round(quantity * unit_price)
        vat_result = vat_module.calculate(amount=line_net, category=category)
        computed_items.append({
            "description": item["description"],
            "quantity": quantity,
            "unit_price": unit_price,
            "category": category,
            "vat_rate": vat_result["rate"],
            "vat_amount": vat_result["vat"],
            "line_net": line_net,
            "line_total": bankers_round(line_net + vat_result["vat"]),
        })

    subtotal = bankers_round(sum(it["line_net"] for it in computed_items))
    total_vat = bankers_round(sum(it["vat_amount"] for it in computed_items))
    total = bankers_round(subtotal + total_vat)

    breakdown_map: dict[float, dict[str, Any]] = {}
    for it in computed_items:
        rate = it["vat_rate"]
        if rate in breakdown_map:
            breakdown_map[rate]["taxable_amount"] = bankers_round(
                breakdown_map[rate]["taxable_amount"] + it["line_net"]
            )
            breakdown_map[rate]["vat_amount"] = bankers_round(
                breakdown_map[rate]["vat_amount"] + it["vat_amount"]
            )
        else:
            vr = vat_module.calculate(amount=100, category=it["category"])
            breakdown_map[rate] = {
                "rate_type": vr["rate_type"],
                "taxable_amount": it["line_net"],
                "vat_amount": it["vat_amount"],
            }

    vat_breakdown = [
        {"rate": rate, "rate_type": d["rate_type"], "taxable_amount": d["taxable_amount"], "vat_amount": d["vat_amount"]}
        for rate, d in breakdown_map.items()
    ]

    inv: dict[str, Any] = {
        "type": type,
        "invoice_number": invoice_number,
        "issue_date": issue_date,
        "due_date": due_date,
        "currency": currency,
        "seller": seller,
        "buyer": buyer,
        "items": computed_items,
        "subtotal": subtotal,
        "total_vat": total_vat,
        "total": total,
        "vat_breakdown": vat_breakdown,
        "purchase_order_ref": purchase_order_ref,
        "notes": notes,
        "validation": {"valid": True, "errors": []},
        "ubl_field_count": 0,
    }

    inv["validation"] = validate(inv)
    inv["ubl_field_count"] = _count_ubl_fields(inv)
    return inv


def validate(inv: dict[str, Any]) -> dict[str, Any]:
    """Validate invoice against UBL 3.0 mandatory fields."""
    errors: list[dict[str, str]] = []

    if not inv.get("invoice_number", "").strip():
        errors.append({"field": "invoice_number", "message": "Invoice number is required"})
    if not inv.get("issue_date", "").strip():
        errors.append({"field": "issue_date", "message": "Issue date is required"})
    if not inv.get("currency", "").strip():
        errors.append({"field": "currency", "message": "Currency code is required"})
    if inv.get("type") not in ("invoice", "credit-note", "debit-note"):
        errors.append({"field": "type", "message": "Invoice type must be invoice, credit-note, or debit-note"})

    seller = inv.get("seller")
    if not seller:
        errors.append({"field": "seller", "message": "Seller details are required"})
    else:
        if not seller.get("name", "").strip():
            errors.append({"field": "seller.name", "message": "Seller name is required"})
        if not seller.get("tin", "").strip():
            errors.append({"field": "seller.tin", "message": "Seller TIN is required"})
        if not seller.get("address", "").strip():
            errors.append({"field": "seller.address", "message": "Seller address is required"})

    buyer = inv.get("buyer")
    if not buyer:
        errors.append({"field": "buyer", "message": "Buyer details are required"})
    else:
        if not buyer.get("name", "").strip():
            errors.append({"field": "buyer.name", "message": "Buyer name is required"})
        if not buyer.get("tin", "").strip():
            errors.append({"field": "buyer.tin", "message": "Buyer TIN is required"})
        if not buyer.get("address", "").strip():
            errors.append({"field": "buyer.address", "message": "Buyer address is required"})

    items = inv.get("items", [])
    if not items:
        errors.append({"field": "items", "message": "At least one line item is required"})
    else:
        for i, item in enumerate(items):
            prefix = f"items[{i}]"
            if not item.get("description", "").strip():
                errors.append({"field": f"{prefix}.description", "message": f"Line item {i + 1} description is required"})
            if item.get("quantity") is None or item["quantity"] <= 0:
                errors.append({"field": f"{prefix}.quantity", "message": f"Line item {i + 1} quantity must be positive"})
            if item.get("unit_price") is None or item["unit_price"] < 0:
                errors.append({"field": f"{prefix}.unit_price", "message": f"Line item {i + 1} unit price must be non-negative"})
            if item.get("vat_rate") is None:
                errors.append({"field": f"{prefix}.vat_rate", "message": f"Line item {i + 1} VAT rate is required"})
            if item.get("vat_amount") is None:
                errors.append({"field": f"{prefix}.vat_amount", "message": f"Line item {i + 1} VAT amount is required"})
            if item.get("line_net") is None:
                errors.append({"field": f"{prefix}.line_net", "message": f"Line item {i + 1} line net amount is required"})
            if item.get("line_total") is None:
                errors.append({"field": f"{prefix}.line_total", "message": f"Line item {i + 1} line total is required"})

    if inv.get("subtotal") is None:
        errors.append({"field": "subtotal", "message": "Subtotal is required"})
    if inv.get("total_vat") is None:
        errors.append({"field": "total_vat", "message": "Total VAT is required"})
    if inv.get("total") is None:
        errors.append({"field": "total", "message": "Total is required"})

    vat_breakdown = inv.get("vat_breakdown", [])
    if not vat_breakdown:
        errors.append({"field": "vat_breakdown", "message": "VAT breakdown is required"})
    else:
        for i, entry in enumerate(vat_breakdown):
            prefix = f"vat_breakdown[{i}]"
            if entry.get("rate") is None:
                errors.append({"field": f"{prefix}.rate", "message": f"VAT breakdown entry {i + 1} rate is required"})
            if entry.get("rate_type") is None:
                errors.append({"field": f"{prefix}.rate_type", "message": f"VAT breakdown entry {i + 1} rate type is required"})
            if entry.get("taxable_amount") is None:
                errors.append({"field": f"{prefix}.taxable_amount", "message": f"VAT breakdown entry {i + 1} taxable amount is required"})
            if entry.get("vat_amount") is None:
                errors.append({"field": f"{prefix}.vat_amount", "message": f"VAT breakdown entry {i + 1} VAT amount is required"})

    return {"valid": len(errors) == 0, "errors": errors}


def to_firs_json(inv: dict[str, Any]) -> str:
    """Generate FIRS-compatible JSON. Deterministic (sorted keys)."""
    obj = {
        "buyer": {
            "address": inv["buyer"]["address"],
            "name": inv["buyer"]["name"],
            "tin": inv["buyer"]["tin"],
            "vrn": inv["buyer"].get("vrn"),
        },
        "currency": inv["currency"],
        "dueDate": inv.get("due_date"),
        "invoiceNumber": inv["invoice_number"],
        "invoiceType": inv["type"],
        "issueDate": inv["issue_date"],
        "items": [
            {
                "category": it.get("category", "standard"),
                "description": it["description"],
                "lineId": i + 1,
                "lineNet": it["line_net"],
                "lineTotal": it["line_total"],
                "quantity": it["quantity"],
                "unitPrice": it["unit_price"],
                "vatAmount": it["vat_amount"],
                "vatRate": it["vat_rate"],
            }
            for i, it in enumerate(inv["items"])
        ],
        "notes": inv.get("notes"),
        "purchaseOrderRef": inv.get("purchase_order_ref"),
        "seller": {
            "address": inv["seller"]["address"],
            "name": inv["seller"]["name"],
            "tin": inv["seller"]["tin"],
            "vrn": inv["seller"].get("vrn"),
        },
        "subtotal": inv["subtotal"],
        "total": inv["total"],
        "totalVat": inv["total_vat"],
        "ublFieldCount": inv["ubl_field_count"],
        "validation": {
            "errors": inv["validation"]["errors"],
            "valid": inv["validation"]["valid"],
        },
        "vatBreakdown": [
            {
                "rate": e["rate"],
                "rateType": e["rate_type"],
                "taxableAmount": e["taxable_amount"],
                "vatAmount": e["vat_amount"],
            }
            for e in inv["vat_breakdown"]
        ],
    }
    return json.dumps(obj, indent=2, sort_keys=True)


def to_csv(invoices: list[dict[str, Any]]) -> str:
    """Generate CSV string with one row per invoice."""
    header = "Invoice Number,Date,Buyer,Seller,Subtotal,VAT,Total,Status"
    rows: list[str] = []
    for inv in invoices:
        status = "valid" if inv["validation"]["valid"] else "invalid"
        rows.append(",".join([
            _csv_escape(inv["invoice_number"]),
            _csv_escape(inv["issue_date"]),
            _csv_escape(inv["buyer"]["name"]),
            _csv_escape(inv["seller"]["name"]),
            f'{inv["subtotal"]:.2f}',
            f'{inv["total_vat"]:.2f}',
            f'{inv["total"]:.2f}',
            status,
        ]))
    return "\n".join([header, *rows])


def _csv_escape(value: str) -> str:
    if "," in value or '"' in value or "\n" in value:
        return f'"{value.replace(chr(34), chr(34) + chr(34))}"'
    return value


def _count_ubl_fields(inv: dict[str, Any]) -> int:
    count = 0
    for key in ("invoice_number", "issue_date", "currency", "type", "due_date", "purchase_order_ref", "notes"):
        if inv.get(key):
            count += 1
    for key in ("subtotal", "total_vat", "total"):
        if inv.get(key) is not None:
            count += 1

    for party_key in ("seller", "buyer"):
        party = inv.get(party_key, {})
        for key in ("name", "tin", "address", "vrn"):
            if party.get(key):
                count += 1

    for item in inv.get("items", []):
        for key in ("description", "quantity", "unit_price", "vat_rate", "vat_amount", "line_net", "line_total"):
            if item.get(key) is not None:
                count += 1

    for entry in inv.get("vat_breakdown", []):
        for key in ("rate", "rate_type", "taxable_amount", "vat_amount"):
            if entry.get(key) is not None:
                count += 1

    return count

File: packages/typescript/src/invoice.ts
// ─── Invoice Module (Layer 2) ────────────────────────────────────────────────
// Creates, validates, and manages invoices with auto-VAT calculation per line item.
// Supports mixed VAT categories, UBL 3.0 mandatory field validation.

import { vat, InvalidAmountError, type TaxCategory } from '@ngtaxkit/core';
import { NgtaxkitError, ErrorCode } from '@ngtaxkit/core';

// ─── Invoice-Specific Errors ─────────────────────────────────────────────────

/** Thrown when a line item has zero or negative quantity. */
export class InvalidQuantityError extends NgtaxkitError {
  constructor(message: string) {
    super('NGTK_INVALID_QUANTITY', message);
    this.name = 'InvalidQuantityError';
  }
}

/** Thrown when an invoice has no line items. */
export class EmptyInvoiceError extends NgtaxkitError {
  constructor(message: string) {
    super('NGTK_EMPTY_INVOICE', message);
    this.name = 'EmptyInvoiceError';
  }
}

// ─── Types ───────────────────────────────────────────────────────────────────

/** Party (seller or buyer) on an invoice. */
export interface Party {
  name: string;
  tin: string;
  address: string;
  vrn?: string;
}

/** A single line item on an invoice. */
export interface InvoiceItem {
  description: string;
  quantity: number;
  unitPrice: number;
  category?: TaxCategory;
}

/** A computed line item with VAT details. */
export interface ComputedInvoiceItem extends InvoiceItem {
  vatRate: number;
  vatAmount: number;
  lineTotal: number;
  lineNet: number;
}

/** VAT breakdown entry grouped by rate. */
export interface VatBreakdown {
  rate: number;
  rateType: 'standard' | 'zero-rated' | 'exempt';
  taxableAmount: number;
  vatAmount: number;
}

/** Validation result for UBL 3.0 mandatory fields. */
export interface ValidationResult {
  valid: boolean;
  errors: ValidationFieldError[];
}

export interface ValidationFieldError {
  field: string;
  message: string;
}

export type InvoiceType = 'invoice' | 'credit-note' | 'debit-note';

/** Options for creating an invoice. */
export interface InvoiceCreateOptions {
  seller: Party;
  buyer: Party;
  items: InvoiceItem[];
  currency?: string;
  issueDate?: string;
  dueDate?: string;
  invoiceNumber: string;
  purchaseOrderRef?: string;
  notes?: string;
  type?: InvoiceType;
}

/** A fully computed invoice. */
export interface Invoice {
  type: InvoiceType;
  invoiceNumber: string;
  issueDate: string;
  dueDate?: string;
  currency: string;
  seller: Party;
  buyer: Party;
  items: ComputedInvoiceItem[];
  subtotal: number;
  totalVat: number;
  total: number;
  vatBreakdown: VatBreakdown[];
  purchaseOrderRef?: string;
  notes?: string;
  validation: ValidationResult;
  ublFieldCount: number;
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

function bankersRound2(value: number): number {
  const shifted = value * 100;
  const floored = Math.floor(shifted);
  const remainder = shifted - floored;
  const epsilon = 1e-9;
  if (Math.abs(remainder - 0.5) < epsilon) {
    return (floored % 2 === 0 ? floored : floored + 1) / 100;
  }
  return Math.round(shifted) / 100;
}

function todayISO(): string {
  const d = new Date();
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

// ─── Create ──────────────────────────────────────────────────────────────────

/**
 * Create an Invoice from options.
 * Auto-calculates VAT per line item, computes subtotal/totalVat/total,
 * builds vatBreakdown grouped by rate, and validates UBL 3.0 mandatory fields.
 */
export function create(options: InvoiceCreateOptions): Invoice {
  const {
    seller,
    buyer,
    items,
    currency = 'NGN',
    issueDate = todayISO(),
    dueDate,
    invoiceNumber,
    purchaseOrderRef,
    notes,
    type = 'invoice',
  } = options;

  // ── Validate inputs ──
  if (!items || items.length === 0) {
    throw new EmptyInvoiceError('Invoice must have at least one line item');
  }

  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    if (item.unitPrice < 0) {
      throw new InvalidAmountError(
        `Line item ${i + 1} has negative unit price: ${item.unitPrice}`,
      );
    }
    if (item.quantity <= 0) {
      throw new InvalidQuantityError(
        `Line item ${i + 1} has zero or negative quantity: ${item.quantity}`,
      );
    }
  }

  // ── Compute line items ──
  const computedItems: ComputedInvoiceItem[] = items.map((item) => {
    const category = item.category ?? 'standard';
    const lineNet = bankersRound2(item.quantity * item.unitPrice);
    const vatResult = vat.calculate({ amount: lineNet, category });
    return {
      ...item,
      category,
      vatRate: vatResult.rate,
      vatAmount: vatResult.vat,
      lineNet,
      lineTotal: bankersRound2(lineNet + vatResult.vat),
    };
  });

  // ── Compute totals ──
  const subtotal = bankersRound2(
    computedItems.reduce((sum, item) => sum + item.lineNet, 0),
  );
  const totalVat = bankersRound2(
    computedItems.reduce((sum, item) => sum + item.vatAmount, 0),
  );
  const total = bankersRound2(subtotal + totalVat);

  // ── Build VAT breakdown grouped by rate ──
  const breakdownMap = new Map<number, { rateType: 'standard' | 'zero-rated' | 'exempt'; taxableAmount: number; vatAmount: number }>();
  for (const item of computedItems) {
    const existing = breakdownMap.get(item.vatRate);
    if (existing) {
      existing.taxableAmount = bankersRound2(existing.taxableAmount + item.lineNet);
      existing.vatAmount = bankersRound2(existing.vatAmount + item.vatAmount);
    } else {
      const vatResult = vat.calculate({ amount: 100, category: item.category ?? 'standard' });
      breakdownMap.set(item.vatRate, {
        rateType: vatResult.rateType,
        taxableAmount: item.lineNet,
        vatAmount: item.vatAmount,
      });
    }
  }

  const vatBreakdown: VatBreakdown[] = Array.from(breakdownMap.entries()).map(
    ([rate, data]) => ({
      rate,
      rateType: data.rateType,
      taxableAmount: data.taxableAmount,
      vatAmount: data.vatAmount,
    }),
  );

  // ── Build invoice ──
  const invoice: Invoice = {
    type,
    invoiceNumber,
    issueDate,
    dueDate,
    currency,
    seller,
    buyer,
    items: computedItems,
    subtotal,
    totalVat,
    total,
    vatBreakdown,
    purchaseOrderRef,
    notes,
    validation: { valid: true, errors: [] },
    ublFieldCount: 0,
  };

  // ── Validate and set field count ──
  const validationResult = validate(invoice);
  invoice.validation = validationResult;
  invoice.ublFieldCount = countUblFields(invoice);

  return invoice;
}

// ─── Validate ────────────────────────────────────────────────────────────────

/**
 * Validate an invoice against UBL 3.0 mandatory field requirements.
 * Returns a ValidationResult with valid=true when all mandatory fields are present,
 * or valid=false with an errors array listing missing/invalid fields.
 *
 * The 55 UBL 3.0 BIS Billing mandatory fields include:
 * - Invoice-level: invoiceNumber, issueDate, currency, type, seller, buyer, items, tax totals
 * - Seller: name, tin, address
 * - Buyer: name, tin, address
 * - Per line item: description, quantity, unitPrice, lineNet, vatAmount, lineTotal, vatRate
 * - Tax totals: subtotal, totalVat, total, vatBreakdown entries
 */
export function validate(invoice: Invoice): ValidationResult {
  const errors: ValidationFieldError[] = [];

  // ── Invoice-level mandatory fields ──
  if (!invoice.invoiceNumber || invoice.invoiceNumber.trim() === '') {
    errors.push({ field: 'invoiceNumber', message: 'Invoice number is required' });
  }
  if (!invoice.issueDate || invoice.issueDate.trim() === '') {
    errors.push({ field: 'issueDate', message: 'Issue date is required' });
  }
  if (!invoice.currency || invoice.currency.trim() === '') {
    errors.push({ field: 'currency', message: 'Currency code is required' });
  }
  if (!invoice.type || !['invoice', 'credit-note', 'debit-note'].includes(invoice.type)) {
    errors.push({ field: 'type', message: 'Invoice type must be invoice, credit-note, or debit-note' });
  }

  // ── Seller mandatory fields ──
  if (!invoice.seller) {
    errors.push({ field: 'seller', message: 'Seller details are required' });
  } else {
    if (!invoice.seller.name || invoice.seller.name.trim() === '') {
      errors.push({ field: 'seller.name', message: 'Seller name is required' });
    }
    if (!invoice.seller.tin || invoice.seller.tin.trim() === '') {
      errors.push({ field: 'seller.tin', message: 'Seller TIN is required' });
    }
    if (!invoice.seller.address || invoice.seller.address.trim() === '') {
      errors.push({ field: 'seller.address', message: 'Seller address is required' });
    }
  }

  // ── Buyer mandatory fields ──
  if (!invoice.buyer) {
    errors.push({ field: 'buyer', message: 'Buyer details are required' });
  } else {
    if (!invoice.buyer.name || invoice.buyer.name.trim() === '') {
      errors.push({ field: 'buyer.name', message: 'Buyer name is required' });
    }
    if (!invoice.buyer.tin || invoice.buyer.tin.trim() === '') {
      errors.push({ field: 'buyer.tin', message: 'Buyer TIN is required' });
    }
    if (!invoice.buyer.address || invoice.buyer.address.trim() === '') {
      errors.push({ field: 'buyer.address', message: 'Buyer address is required' });
    }
  }

  // ── Items mandatory ──
  if (!invoice.items || invoice.items.length === 0) {
    errors.push({ field: 'items', message: 'At least one line item is required' });
  } else {
    for (let i = 0; i < invoice.items.length; i++) {
      const item = invoice.items[i];
      const prefix = `items[${i}]`;

      if (!item.description || item.description.trim() === '') {
        errors.push({ field: `${prefix}.description`, message: `Line item ${i + 1} description is required` });
      }
      if (item.quantity == null || item.quantity <= 0) {
        errors.push({ field: `${prefix}.quantity`, message: `Line item ${i + 1} quantity must be positive` });
      }
      if (item.unitPrice == null || item.unitPrice < 0) {
        errors.push({ field: `${prefix}.unitPrice`, message: `Line item ${i + 1} unit price must be non-negative` });
      }
      if (item.vatRate == null) {
        errors.push({ field: `${prefix}.vatRate`, message: `Line item ${i + 1} VAT rate is required` });
      }
      if (item.vatAmount == null) {
        errors.push({ field: `${prefix}.vatAmount`, message: `Line item ${i + 1} VAT amount is required` });
      }
      if (item.lineNet == null) {
        errors.push({ field: `${prefix}.lineNet`, message: `Line item ${i + 1} line net amount is required` });
      }
      if (item.lineTotal == null) {
        errors.push({ field: `${prefix}.lineTotal`, message: `Line item ${i + 1} line total is required` });
      }
    }
  }

  // ── Tax totals mandatory fields ──
  if (invoice.subtotal == null) {
    errors.push({ field: 'subtotal', message: 'Subtotal is required' });
  }
  if (invoice.totalVat == null) {
    errors.push({ field: 'totalVat', message: 'Total VAT is required' });
  }
  if (invoice.total == null) {
    errors.push({ field: 'total', message: 'Total is required' });
  }

  // ── VAT breakdown mandatory ──
  if (!invoice.vatBreakdown || invoice.vatBreakdown.length === 0) {
    errors.push({ field: 'vatBreakdown', message: 'VAT breakdown is required' });
  } else {
    for (let i = 0; i < invoice.vatBreakdown.length; i++) {
      const entry = invoice.vatBreakdown[i];
      const prefix = `vatBreakdown[${i}]`;
      if (entry.rate == null) {
        errors.push({ field: `${prefix}.rate`, message: `VAT breakdown entry ${i + 1} rate is required` });
      }
      if (entry.rateType == null) {
        errors.push({ field: `${prefix}.rateType`, message: `VAT breakdown entry ${i + 1} rate type is required` });
      }
      if (entry.taxableAmount == null) {
        errors.push({ field: `${prefix}.taxableAmount`, message: `VAT breakdown entry ${i + 1} taxable amount is required` });
      }
      if (entry.vatAmount == null) {
        errors.push({ field: `${prefix}.vatAmount`, message: `VAT breakdown entry ${i + 1} VAT amount is required` });
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

// ─── UBL Field Count ─────────────────────────────────────────────────────────

/**
 * Count the number of UBL 3.0 mandatory fields present in the invoice.
 * The 55 mandatory fields are distributed across invoice-level, seller, buyer,
 * line items, and tax summary sections.
 */
function countUblFields(invoice: Invoice): number {
  let count = 0;

  // Invoice-level fields (10)
  if (invoice.invoiceNumber) count++;
  if (invoice.issueDate) count++;
  if (invoice.currency) count++;
  if (invoice.type) count++;
  if (invoice.subtotal != null) count++;
  if (invoice.totalVat != null) count++;
  if (invoice.total != null) count++;
  if (invoice.dueDate) count++;
  if (invoice.purchaseOrderRef) count++;
  if (invoice.notes) count++;

  // Seller fields (4)
  if (invoice.seller?.name) count++;
  if (invoice.seller?.tin) count++;
  if (invoice.seller?.address) count++;
  if (invoice.seller?.vrn) count++;

  // Buyer fields (4)
  if (invoice.buyer?.name) count++;
  if (invoice.buyer?.tin) count++;
  if (invoice.buyer?.address) count++;
  if (invoice.buyer?.vrn) count++;

  // Per line item fields (7 each)
  for (const item of invoice.items) {
    if (item.description) count++;
    if (item.quantity != null) count++;
    if (item.unitPrice != null) count++;
    if (item.vatRate != null) count++;
    if (item.vatAmount != null) count++;
    if (item.lineNet != null) count++;
    if (item.lineTotal != null) count++;
  }

  // VAT breakdown fields (4 each)
  for (const entry of invoice.vatBreakdown) {
    if (entry.rate != null) count++;
    if (entry.rateType) count++;
    if (entry.taxableAmount != null) count++;
    if (entry.vatAmount != null) count++;
  }

  return count;
}

// ─── FIRS JSON Generation ────────────────────────────────────────────────────

/**
 * Generate FIRSMBS-compatible JSON string with all 55 mandatory fields.
 * Deterministic: identical input produces identical JSON output.
 * Keys are sorted to ensure determinism.
 */
export function toFIRSJSON(invoice: Invoice): string {
  const obj = {
    invoiceNumber: invoice.invoiceNumber,
    invoiceType: invoice.type,
    issueDate: invoice.issueDate,
    dueDate: invoice.dueDate ?? null,
    currency: invoice.currency,
    purchaseOrderRef: invoice.purchaseOrderRef ?? null,
    notes: invoice.notes ?? null,
    seller: {
      name: invoice.seller.name,
      tin: invoice.seller.tin,
      address: invoice.seller.address,
      vrn: invoice.seller.vrn ?? null,
    },
    buyer: {
      name: invoice.buyer.name,
      tin: invoice.buyer.tin,
      address: invoice.buyer.address,
      vrn: invoice.buyer.vrn ?? null,
    },
    items: invoice.items.map((item, i) => ({
      lineId: i + 1,
      description: item.description,
      quantity: item.quantity,
      unitPrice: item.unitPrice,
      category: item.category ?? 'standard',
      vatRate: item.vatRate,
      vatAmount: item.vatAmount,
      lineNet: item.lineNet,
      lineTotal: item.lineTotal,
    })),
    subtotal: invoice.subtotal,
    totalVat: invoice.totalVat,
    total: invoice.total,
    vatBreakdown: invoice.vatBreakdown.map((entry) => ({
      rate: entry.rate,
      rateType: entry.rateType,
      taxableAmount: entry.taxableAmount,
      vatAmount: entry.vatAmount,
    })),
    validation: {
      valid: invoice.validation.valid,
      errors: invoice.validation.errors,
    },
    ublFieldCount: invoice.ublFieldCount,
  };
  return JSON.stringify(obj, null, 2);
}

// ─── CSV Export ──────────────────────────────────────────────────────────────

/**
 * Generate a CSV string with one row per invoice.
 * Columns: Invoice Number, Date, Buyer, Seller, Subtotal, VAT, Total, Status
 */
export function toCSV(invoices: Invoice[]): string {
  const header = 'Invoice Number,Date,Buyer,Seller,Subtotal,VAT,Total,Status';
  const rows = invoices.map((inv) => {
    const status = inv.validation.valid ? 'valid' : 'invalid';
    return [
      csvEscape(inv.invoiceNumber),
      csvEscape(inv.issueDate),
      csvEscape(inv.buyer.name),
      csvEscape(inv.seller.name),
      inv.subtotal.toFixed(2),
      inv.totalVat.toFixed(2),
      inv.total.toFixed(2),
      status,
    ].join(',');
  });
  return [header, ...rows].join('\n');
}

/** Escape a CSV field: wrap in quotes if it contains comma, quote, or newline. */
function csvEscape(value: string): string {
  if (value.includes(',') || value.includes('"') || value.includes('\n')) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}
